src/zope/traversing/namespace.py

```diff
@@ -673,7 +673,7 @@ def traverse(self, name, ignored):
                 # if we want to enable tracebacks when also trying to
                 # debug a different skin?
                 skin = zope.component.getUtility(IBrowserSkinType, 'Debug')
-                directlyProvides(request, providedBy(request) + skin)
+                directlyProvides(request, skin)
```

Comment:
Uhh, this leaves me worried.

I remember there being an `alsoProvides()`, which means `directlyProvides()` removes the old things that were directly provided (such as the skin), instead of adding to them.

++debug++errors should not be removing old skins.

Reply:
The old code might've been buggy and should've been using `alsoProvides()` (which expands to `directlyProvides(request, directlyProvidedBy(request), skin)`).

The test suite might also be missing a test, that this traversal doesn't remove any custom interfaces directlyProvided() on the request instance.

(I have an application relying on this kind of thing.)

Reply:
You're exactly right, thanks for the catch! 

It seems backwards that zope.interface does it that way. It seems like the (hypothetical) more complicated spelling `directlyProvidesOnly` should be what erases anything already directly provided.

Reply:
Yeah, I believe it was a design mistake, never fixed because of backwards compatibility concerns.